src/openrct2-ui/windows/TrackDesignPlace.cpp

```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
+            if (res.Error == GameActions::Status::Ok || res.Error == GameActions::Status::InsufficientFunds)
             {
-                uint8_t* pixel = DrawMiniPreviewGetPixelPtr(pixelPosition);
-                uint8_t colour = entrance.isExit ? _PaletteIndexColourExit : _PaletteIndexColourEntrance;
-                for (int32_t i = 0; i < 4; i++)
-                {
-                    pixel[338 + i] = colour; // x + 2, y + 2
-                    pixel[168 + i] = colour; //        y + 1
-                    pixel[2 + i] = colour;   // x + 2
-                    pixel[172 + i] = colour; // x + 4, y + 1
-                }
+                return res;
             }
         }
+        return res;
     }
-}
+};
 
-static void WindowTrackPlaceDrawMiniPreviewMaze(
-    TrackDesign* td6, int32_t pass, const CoordsXY& origin, CoordsXY min, CoordsXY max)
+rct_window* WindowTrackPlaceOpen(const track_design_file_ref* tdFileRef)
 {
-    uint8_t rotation = (_currentTrackPieceDirection + get_current_rotation()) & 3;
-    for (const auto& mazeElement : td6->maze_elements)
+    _trackDesign = TrackDesignImport(tdFileRef->path);
+    if (_trackDesign == nullptr)
     {
-        auto rotatedMazeCoords = origin + TileCoordsXY{ mazeElement.x, mazeElement.y }.ToCoordsXY().Rotate(rotation);
-
-        if (pass == 0)
-        {
-            min.x = std::min(min.x, rotatedMazeCoords.x);
-            max.x = std::max(max.x, rotatedMazeCoords.x);
-            min.y = std::min(min.y, rotatedMazeCoords.y);
-            max.y = std::max(max.y, rotatedMazeCoords.y);
-        }
-        else
-        {
-            auto pixelPosition = DrawMiniPreviewGetPixelPosition(rotatedMazeCoords);
-            if (DrawMiniPreviewIsPixelInBounds(pixelPosition))
-            {
-                uint8_t* pixel = DrawMiniPreviewGetPixelPtr(pixelPosition);
+        return nullptr;
+    }
 
-                uint8_t colour = _PaletteIndexColourTrack;
+    window_close_construction_windows();
 
-                // Draw entrance and exit with different colours.
-                if (mazeElement.type == MAZE_ELEMENT_TYPE_ENTRANCE)
-                    colour = _PaletteIndexColourEntrance;
-                else if (mazeElement.type == MAZE_ELEMENT_TYPE_EXIT)
-                    colour = _PaletteIndexColourExit;
+    _window_track_place_mini_preview.resize(TRACK_MINI_PREVIEW_SIZE);
 
-                for (int32_t i = 0; i < 4; i++)
-                {
-                    pixel[338 + i] = colour; // x + 2, y + 2
-                    pixel[168 + i] = colour; //        y + 1
-                    pixel[2 + i] = colour;   // x + 2
-                    pixel[172 + i] = colour; // x + 4, y + 1
-                }
-            }
-        }
+    auto* window = window_bring_to_front_by_class(WC_TRACK_DESIGN_PLACE);
+    if (window == nullptr)
+    {
+        window = WindowCreate<TrackDesignPlaceWindow>(WC_TRACK_DESIGN_PLACE, WW, WH, 0);
```

Comment:
There is a `WindowFocusOrCreate` that you can use to handle the window_bring_to_front_by_class all as a oner.